### PR TITLE
Manual grading

### PR DESCRIPTION
**Pull request by frostyfan109**

### CODE REVIEW

File: app/api/api_v1/endpoints/assignment_router.py
Comment (Hoid) [inline app/api/api_v1/endpoints/assignment_router.py:41]: What is this proportion? Does it mean how much the grade is weighted in the class towards the final grade, or?

Comment (frostyfan109) [reply app/api/api_v1/endpoints/assignment_router.py:41]: It's the student's grade for the submission as a proportion

Comment (frostyfan109) [reply app/api/api_v1/endpoints/assignment_router.py:41]: It was renamed from grade percent after harit misinterpreted grade percent to be in [0,100].

Comment (Hoid) [reply app/api/api_v1/endpoints/assignment_router.py:41]: Oohh this is manual grading, that's right, my bad. Okay this makes sense.

File: app/api/api_v1/endpoints/assignment_router.py
Comment (Hoid) [inline app/api/api_v1/endpoints/assignment_router.py]: This should be named `grade_assignment_manually()` not just `grade_assignment()` since the endpoint is `.../grade_manual`

File: app/services/submission_service.py
Comment (Hoid) [inline app/services/submission_service.py]: I'm worried that if there's an error thrown in the stuff you've added the stuff done previously needs to be rolled back. Maybe move this part to below the code you added?

```
self.session.add(submission)
self.session.commit()
```

Comment (frostyfan109) [reply app/services/submission_service.py]: this change is done in submission downsync. I'm not particularly sure how the syncing change above got included in manual-grading and submission-downsync but not in submission-upload-cleanup which is the parent of both...

File: app/core/exceptions/grading.py
Comment (Hoid) [inline app/core/exceptions/grading.py]: I would use a term here other than "batch grading" if you expect that this error might ever be shown to users. It's kind of an esoteric term for the lay person, and the message was kind of confusing for me too at first. What about "only one submission per student is permitted during grading"? We don't have to stipulate "batch grading" at all since the scenario where we grade only one submission total precludes the possibility of attempting to grade multiple submissions from one student.

File: app/services/canvas_service.py
Comment (Hoid) [inline app/services/canvas_service.py:385]: The comment above this says it can be any value between 0 and 1. Why not enforce that with a type? Maybe like this?
```python
class GradeProportion(float):
    def __new__(cls, input = 0, *args, **kwargs):
        instance = int.__new__(cls, input, *args, **kwargs)
        if not 0 <= instance <= 1:
            raise ValueError('Value outside GradeProportion range')
        return instance
```

Comment (Hoid) [reply app/services/canvas_service.py:385]: Then you can only accept a `GradeProportion` here instead of any float, and you won't have to add validation within this method.

Comment (frostyfan109) [reply app/services/canvas_service.py:385]: Oh yes there's a way to do that with pydantic I think

Comment (frostyfan109) [reply app/services/canvas_service.py:385]: It's confloat with ge=0, le=1

Comment (Hoid) [reply app/services/canvas_service.py:385]: Looks like its use is discouraged in the docs, since it doesn't play well with static analysis tools: https://docs.pydantic.dev/latest/api/types/#pydantic.types.confloat

They give another solution, to use this
```python
class Foo(BaseModel):
    bar: Annotated[float, Field(strict=True, gt=0)]
```
which I thiiink works just as well as the class I gave above. I do like that the class implementation is named, though, rather than the field just having properties that you don't know of in the code unless you look on the base model.

Comment (frostyfan109) [reply app/services/canvas_service.py:385]: An annotated float works just as well. I would keep in mind we are using I believe pydantic v1.

File: app/services/grading_service.py
Comment (Hoid) [inline app/services/grading_service.py]: Can this be named `grade_assignment_manually()`? lol If you care to keep it the same it's fine it's just a preference I have that you might agree with 🤷‍♂️

Comment (frostyfan109) [reply app/services/grading_service.py]: Yeah I don't particularly care

File: app/services/canvas_service.py
Comment (Hoid) [reply app/services/canvas_service.py:385]: Sounds good. So are you going to change this float to an annotated float with constraints, or leave it as is?

File: app/services/lms_sync_service.py
Comment (Hoid) [inline app/services/lms_sync_service.py:203]: Is this just done as validation? Don't we already have the student's information under `submission.student`?

Comment (frostyfan109) [reply app/services/lms_sync_service.py:203]: that's the canvas student, not the database student. i named it as such to keep in line with elsewhere in the service.

File: app/services/canvas_service.py
Comment (frostyfan109) [reply app/services/canvas_service.py:385]: yes i will test it

File: app/services/submission_service.py
Comment (Hoid) [reply app/services/submission_service.py]: So...what needs to happen here? Can we make this code exactly the way it should be for now?

File: app/services/lms_sync_service.py
Comment (Hoid) [reply app/services/lms_sync_service.py:203]: I think it'd be better to just change them all to be named `canvas_student`, which is more clear considering there's a database student being used up above in a lot of these methods, instead of just acquiescing to the norm.

**Comment by Hoid** [comment]
Okay I've reviewed all the code and other than the outstanding convos in this PR I think it's all good. Just need to resolve those convos and I think it's good to merge.

File: app/services/submission_service.py
Comment (frostyfan109) [reply app/services/submission_service.py]: This code is outdated, refer to latest version of it.

Comment (frostyfan109) [reply app/services/submission_service.py]: In fairness, even after merging in the initial submission sync, the actual cleanup code seems to live on the second "submission downsync" PR, so the issue still persists it's just been relocated.

File: app/services/lms_sync_service.py
Comment (Hoid) [reply app/services/lms_sync_service.py:203]: Actually nevermind I think this one is fine, don't want to hold up the PR for it

**Review by Hoid — APPROVED** [review]

**Pull request merged**
